### PR TITLE
Sanitizes the keys passed in the payload dictionary to the SanitizingJSONEncoder

### PR DESCRIPTION
**Pull request by brockhaywood**
…JSONEncoder

## Goal
In some circumstances, it may be possible to crash the SanitizingJSONEncoder by passing bad keys in the payload to the SanitizingJSONEncoder. This PR attempts to sanitize the keys to ensure they are all strings.

### Changed
Changed the `_sanitize_dict` function of the `SanitizingJSONEncoder` to ensure that all keys are strings. If a key cannot be converted to a string then it is dropped from the clean dictionary as this will crash with it.

## Tests

Added `test_json_encode_invalid_keys` to `TestUtils`

### Linked issues

Fixes https://github.com/bugsnag/bugsnag-python/issues/162

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: bugsnag/utils.py [hunk 1]
```diff
@@ -118,7 +118,19 @@ def _sanitize_dict(self, obj, trim_strings, ignored):
...
+            if isinstance(key, basestring):
+                clean_dict[key] = clean_value
+            else:
+                try:
+                    clean_dict[str(key)] = clean_value
+                except:
+                    try:
+                        clean_dict[repr(key)] = clean_value
+                    except:
+                        pass
```
Comment (brockhaywood) [inline bugsnag/utils.py]: If we can't sanitize the key then drop it from the clean_dict otherwise we know this will crash.

File: bugsnag/utils.py [hunk 2]
```diff
@@ -118,7 +118,19 @@ def _sanitize_dict(self, obj, trim_strings, ignored):
 
         clean_dict = {}
         for key, value in six.iteritems(obj):
-            clean_dict[key] = self._sanitize(value, trim_strings, ignored)
+
+            clean_value = self._sanitize(value, trim_strings, ignored)
+
+            if isinstance(key, basestring):
```
Comment (Cawllec) [inline bugsnag/utils.py]: Our library is used in both Python 2.7 and 3, as such `six.string_types` is necessary here for compatibility.

File: bugsnag/utils.py [hunk 3]
```diff
@@ -118,7 +118,19 @@ def _sanitize_dict(self, obj, trim_strings, ignored):
...
+                clean_dict[key] = clean_value
+            else:
```
Comment (Cawllec) [inline bugsnag/utils.py]: This may be cleaner split out into a separate function

File: bugsnag/utils.py [hunk 4]
```diff
@@ -118,7 +118,19 @@ def _sanitize_dict(self, obj, trim_strings, ignored):
...
+                try:
+                    clean_dict[str(key)] = clean_value
+                except:
+                    try:
+                        clean_dict[repr(key)] = clean_value
```
Comment (Cawllec) [inline bugsnag/utils.py]: We should be wary of using the `repr` function response as a key, it may produce a lot more information than was intended for use in this area.

File: bugsnag/utils.py (hunk 1 above)
Comment (Cawllec) [inline bugsnag/utils.py]: In the event an object couldn't be converted into a string using either `str` or `repr`, this could cause further problems as data would be being dropped with no indication.  If we could log the dropped data at this point, that may be a better solution.

File: tests/test_utils.py [hunk 5]
```diff
@@ -184,3 +184,18 @@ def test_sanitize_list_handling(self):
...
         # no assert as we are just expecting this not to throw
         encoder._sanitize(data, ['password'], ['password'])
+
+    def test_json_encode_invalid_keys(self):
+        """
+        Test that _sanitize can accept some invalid json where a function
+        name or some other bad data is passed as a key in the payload dictionary.
+        """
+        def foo():
+            return "123"
```
Comment (Cawllec) [inline tests/test_utils.py]: We'll need to expand the number of test cases for other potentially invalid keys.

Comment (brockhaywood) [reply tests/test_utils.py]: Yea, I didn't have any immediate good ideas on other test cases where a key would be invalid but wouldn't be handled by the json encoder. For example, ints and None are handled internally and dont cause exceptions.

**Comment by brockhaywood** [comment]
@Cawllec updated based on your feedback and tests are passing

**Comment by brockhaywood** [comment]
@Cawllec can you re-review?

**Comment by brockhaywood** [comment]
@Cawllec cool. updated the message and rebased against master
